--- /home/kbtch_/Documents/MXSS_HtmlSanitizer_Symphony/dom-php-8.4.18/lexbor/lexbor/html/parser.c	2026-01-27 02:06:20.000000000 +0100
+++ /home/kbtch_/Documents/MXSS_HtmlSanitizer_Symphony/lexbor-2.7.0/source/lexbor/html/parser.c	2026-03-03 11:50:38.000000000 +0100
@@ -13,9 +13,10 @@
 #include "lexbor/html/tree/template_insertion.h"
 #include "lexbor/html/tree/insertion_mode.h"
 
-#define LXB_HTML_TAG_RES_DATA
-#define LXB_HTML_TAG_RES_SHS_DATA
-#include "lexbor/html/tag_res.h"
+#ifndef LEXBOR_DISABLE_INTERNAL_EXTERN
+    LXB_EXTERN lxb_html_tag_category_t lxb_html_tag_res_cats[LXB_TAG__LAST_ENTRY][LXB_NS__LAST_ENTRY];
+    LXB_EXTERN lxb_html_tag_fixname_t lxb_html_tag_res_fixname_svg[LXB_TAG__LAST_ENTRY];
+#endif
 
 
 static void
@@ -43,6 +44,8 @@
         return status;
     }
 
+    lxb_html_tokenizer_keep_duplicate_set(parser->tkz, true);
+
     /* Tree */
     parser->tree = lxb_html_tree_create();
     status = lxb_html_tree_init(parser->tree, parser->tkz);
@@ -339,7 +342,9 @@
         parser->tree->fragment = NULL;
     }
 
-    if (lxb_html_document_is_original(parser->tree->document) == false) {
+    if (parser->tree->document != NULL
+        && lxb_html_document_is_original(parser->tree->document) == false)
+    {
         if (parser->root != NULL) {
             doc = lxb_dom_interface_node(parser->tree->document)->owner_document;
             parser->root->parent = &doc->node;
